### PR TITLE
[SPARK-32550][SQL] Make SpecificInternalRow constructors faster by using while loops instead of maps

### PR DESCRIPTION
**Pull request by msamirkhan**
### What changes were proposed in this pull request?
Change maps in two constructors of SpecificInternalRow to while loops.

### Why are the changes needed?
This was originally noticed with https://github.com/apache/spark/pull/29353 and https://github.com/apache/spark/pull/29354 and will have impacts on performance of reading ORC and Avro files. Ran AvroReadBenchmarks with the new cases of nested and array'd structs in https://github.com/apache/spark/pull/29352. Haven't run benchmarks for ORC but can do that if needed.

**Before:**
```
Nested Struct Scan:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
Nested Struct                                     74674          75319         912          0.0      142429.1       1.0X

Array of Struct Scan:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
Array of Structs                                  34193          34339         206          0.0       65217.9       1.0X
```
**After:**
```
Nested Struct Scan:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
Nested Struct                                     48451          48619         237          0.0       92413.2       1.0X

Array of Struct Scan:                     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
Array of Structs                                  18518          18683         234          0.0       35319.6       1.0X
```

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Ran AvroReadBenchmarks with the new cases of nested and array'd structs in https://github.com/apache/spark/pull/29352.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by dongjoon-hyun** [comment]
Retest this please

**Comment by HyukjinKwon** [comment]
retest this please

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala [hunk 1]
```diff
@@ -192,24 +192,41 @@ final class MutableAny extends MutableValue {
...
-        // We use INT for DATE internally
-        case IntegerType | DateType => new MutableInt
-        // We use Long for Timestamp internally
-        case LongType | TimestampType => new MutableLong
-        case FloatType => new MutableFloat
-        case DoubleType => new MutableDouble
-        case _ => new MutableAny
-      }.toArray)
+  private[this] def dataTypeToMutableValue(dataType: DataType): MutableValue = dataType match {
+    case BooleanType => new MutableBoolean
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:196]: Seems plausible. While micro-optimizing, I wonder if the ordering of these matches matters? is it faster to put more common types earlier?

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:196]: I think this is a good point.

Comment (msamirkhan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:196]: I will take a look.

Comment (msamirkhan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:196]: Did not see a lot of difference. Ran benchmarks on avro format with the following schema and 1024 * 1024 * 15 data points and 10 iterations.
```
root
|-- col1: struct (nullable = true)
|    |-- _1: integer (nullable = false)
|    |-- _2: integer (nullable = false)
```
With Integer type at bottom, getting between 7.1 to 7.7 seconds and with Integer type at top, getting between 7 and 7.5 seconds.

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:196]: OK. Might be worth reordering to integers, floating point, boolean, then the rest, but it's no big deal.

Comment (msamirkhan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:199]: Done

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/SpecificInternalRow.scala:199]: Oh, I mean something more like:

Integer
Long
Float
Double
Boolean
Byte
Short

My guess is that's roughly in order of prevalence.

**Review by srowen — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Comment by LuciferYang** [comment]
@srowen @HyukjinKwon @dongjoon-hyun @msamirkhan hi~ has anyone paid attention to the performance impact of this issue? I found that it had some negative impact on performance and create a new Jira [SPARK-32690](https://issues.apache.org/jira/browse/SPARK-32690) .

the typical cases is "deterministic cardinality estimation" in 

`HyperLogLogPlusPlusSuite` when rsd is 0.001, we found the code that is significantly slower is line 41 in `HyperLogLogPlusPlusSuite`

https://github.com/apache/spark/blob/08b951b1cb58cea2c34703e43202fe7c84725c8a/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlusSuite.scala#L40-L44

The size of "hll.aggBufferAttributes" in this case is 209716, the results of comparison before and after spark-32550 merged are as follows:


  | After   SPARK-32550 create createBuffer | After   SPARK-32550 end to end | Before   SPARK-32550 create input | Before   SPARK-32550 end to end
-- | -- | -- | -- | --
rsd 0.001, n   1000 | 52715513243 | 53004810687 | 195807999 | 773977677
rsd 0.001, n   5000 | 51881246165 | 52519358215 | 13689949 | 249974855
rsd 0.001, n   10000 | 52234282788 | 52374639172 | 14199071 | 183452846
rsd 0.001, n   50000 | 55503517122 | 55664035449 | 15219394 | 584477125
rsd 0.001, n   100000 | 51862662845 | 52116774177 | 19662834 | 166483678
rsd 0.001, n   500000 | 51619226715 | 52183189526 | 178048012 | 16681330
rsd 0.001, n   1000000 | 54861366981 | 54976399142 | 226178708 | 18826340
rsd 0.001, n   5000000 | 52023602143 | 52354615149 | 388173579 | 15446409
rsd 0.001, n   10000000 | 53008591660 | 53601392304 | 533454460 | 16033032

We can use `mvn test -pl sql/catalyst -DwildcardSuites=org.apache.spark.sql.catalyst.expressions.aggregate.HyperLogLogPlusPlusSuite -Dtest=none` to verify the result above

**Comment by LuciferYang** [comment]
@srowen @HyukjinKwon @dongjoon-hyun @msamirkhan I have git a [pr](https://github.com/apache/spark/pull/29529) to revert this, this feature seems to have some bad cases

**Comment by srowen** [comment]
I'll add more comments to your other PR; maybe there is a way to make it all faster. There is a tradeoff here

**Comment by LuciferYang** [comment]
@srowen ok

**Pull request closed without merging**